src/TorchSharp/Utils/ShuffleGenerator.cs

```diff
@@ -0,0 +1,89 @@
+using System;
+
+namespace TorchSharp.Utils
+{
+    public class ShuffleGenerator
+    {
+        int maxrange;
+        int prime;
+        int index;
+        int offset;
+        int runningvalue;
+
+        public ShuffleGenerator(int size)
+        {
+            var min = size / 2;
+            maxrange = size;
+            prime = selectCoPrimeResev(min, size);
+            offset = new Random().Next(size);
+            index = 0;
+            runningvalue = offset;
+        }
+
+        private int getCurrentValue()
+        {
+            return (int) (((long) index * prime + offset) % (maxrange));
+        }
+
+        public bool hasNext()
+        {
+            return index < maxrange;
+        }
+
+        public int next()
+        {
+            runningvalue += prime;
+            if (runningvalue >= maxrange) runningvalue -= maxrange;
+            index++;
+            return runningvalue;
+        }
+
+        private const int MAX_COUNT = int.MaxValue;
+
+        static int selectCoPrimeResev(int min, int target)
+        {
+            var count = 0;
+            var selected = 0;
+            var rand = new Random();
+            for (var val = min; val < target; ++val) {
+                if (coprime(val, target)) {
+                    count += 1;
+                    if ((count == 1) || (rand.Next(count) < 1)) {
+                        selected = val;
+                    }
+                }
+
+                if (count == MAX_COUNT) return val;
+            }
+
+            return selected;
+        }
+
+        static bool coprime(int u, int v) => gcd(u, v) == 1;
+
+        static int gcd(int u, int v)
```

Comment:
Is this Greatest Common Divisor? I don't know much about which is the most efficient way to compute it, but this looks a lot more complex than some of the algorithms I found on stackoverflow:

https://stackoverflow.com/questions/18541832/c-sharp-find-the-greatest-common-divisor